invitev2.go

```diff
@@ -0,0 +1,125 @@
+package gomatrixserverlib
+
+import (
+	"encoding/json"
+	"errors"
+
+	"github.com/tidwall/gjson"
+)
+
+// InviteV2Request and InviteV2StrippedState are defined in
+// https://matrix.org/docs/spec/server_server/r0.1.3#put-matrix-federation-v2-invite-roomid-eventid
+
+func NewInviteV2Request(event *HeaderedEvent, state []InviteV2StrippedState) (
+	request InviteV2Request, err error,
+) {
+	if event.RoomVersion == "" {
+		err = errors.New("gomatrixserverlib: malformed headered event")
+		return
+	}
+	request.fields.inviteV2RequestHeaders = inviteV2RequestHeaders{
+		RoomVersion:     event.RoomVersion,
+		InviteRoomState: state,
+	}
+	request.fields.Event = event.Unwrap()
+	return
+}
+
+type inviteV2RequestHeaders struct {
+	RoomVersion     RoomVersion             `json:"room_version"`
+	InviteRoomState []InviteV2StrippedState `json:"invite_stripped_state"`
+}
+
+// InviteV2Request is used in a /_matrix/federation/v2/invite request.
+type InviteV2Request struct {
+	fields struct {
+		inviteV2RequestHeaders
+		Event Event `json:"event"`
+	}
+}
+
+// MarshalJSON implements json.Marshaller
+func (i InviteV2Request) MarshalJSON() ([]byte, error) {
+	return json.Marshal(i.fields)
+}
+
+// UnmarshalJSON implements json.Unmarshaller
+func (i *InviteV2Request) UnmarshalJSON(data []byte) error {
+	err := json.Unmarshal(data, &i.fields.inviteV2RequestHeaders)
+	if err != nil {
+		return err
+	}
+	eventJSON := gjson.GetBytes(data, "event")
+	if !eventJSON.Exists() {
+		return errors.New("gomatrixserverlib: request doesn't contain event")
+	}
+	i.fields.Event, err = NewEventFromUntrustedJSON([]byte(eventJSON.String()), i.fields.RoomVersion)
```

Comment:
Why do we typecast to string then `[]byte`? Can't we just use `eventJSON` as-is? Also, we don't check if `i.fields.RoomVersion` is sane.

Reply:
Annoyingly this is because `gjson.GetBytes()` returns a `gjson.Result` rather than a `[]byte`, and seemingly none of the functions on `gjson.Result` just return a `[]byte`.